### PR TITLE
[SPARK-30295][SQL] Replacing java code with Scala implementations in the Spark-sql cli

### PR DESCRIPTION
**Pull request by javierivanov**
#### Removed Hive dependencies from:
- [x] SparkSQLCLIDriver
- [x] SparkSQLDriver
#### Features added :
- [x] Added a SparkSQLCLIArguments
- [x] Added new prompt string `spark-sql ($database)>`
- [x] Added same showString method from Datasets to print String formatted results.

Actual Spark-sql client with Dataset show method-like display.
```
+---+-----+-------+
| id| name|address|
+---+-----+-------+
|  5|name3|   add3|
|  5|name1|   add1|
|  5|name2|   add2|
+---+-----+-------+
```

#### TODO:
- [x] Code documentation
- [x] Interruption Handling.
- [x] IO handling for reading and writtig streams, currently using prinln.

### Why are the changes needed?

To remove unnecessary hive dependencies for the Spark SQL Client.

### Does this PR introduce any user-facing change?
 
No

### How was this patch tested?

Added test for: 

> SPARK-30049 Should not complaint for quotes in commented lines

Passing current tests

### CODE REVIEW

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    "--file" -> "file",
+    "-i" -> "init-file",
+    "--database" -> "database",
+    "--hiveconf" -> "hiveconf",
+    "-e" -> "quoted-query-string",
+    "-f" -> "file",
+    "--hivevar" -> "hivevar",
+    "-d" -> "define",
+    "--define" -> "define",
+    "--conf" -> "conf"
```
Comment (seanli-rallyhealth) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala]: can you group keys by alpha beta order?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+  /**
+   * Usage string helper to show options for --help and  illegal arguments.
+   */
+  lazy val usage: String = {
+    var output = "Usage:\n"
+    output += optionsArgMap.map(pair => s"${pair._1} [ARG]\t${pair._2}").mkString("\n")
+    output += "\n\nOptions:\n"
+    output += optionsMap.map(pair => s"-${pair._1}|--${pair._2}\t${pair._2}").mkString("\n")
+    output
```
Comment (seanli-rallyhealth) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala]: use multiple line string?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+      if (innerList.isEmpty) {
+        // End of args list.
+        accumulator
+      } else {
+        if (prevKey.isEmpty) {
+
+          // If current arg is not a key and also
+          // there is no previous key, should throw exception.
+          if (!innerList.head.startsWith("-")) {
```
Comment (seanli-rallyhealth) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala]: use .head is risky of throwing exception if innerList is empty.
suggest to use .headOption.exists(_.startsWith("-"))

Comment (javierivanov) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala]: The head operation is covered with a if (innerList.isEmpty) should be safe.

**Comment by javierivanov** [comment]
Hey @HyukjinKwon , @dongjoon-hyun any thoughts for this PR? Please, let me know. Thanks :)

**Comment by srowen** [comment]
Is the TL;DR here that you are inlining some Hive code so as to not depend on it? while I can see the value of reducing dependencies, does this much help much?

**Comment by javierivanov** [comment]
> Is the TL;DR here that you are inlining some Hive code so as to not depend on it? while I can see the value of reducing dependencies, does this much help much?

@srowen By removing hive dependencies where it is not needed, it is a step forward to reduce the footprint of the code and to maybe have a future version of the thrift module without Hive deps. at all. It also adds maintainability since Java collections were replaced with scala ones and some procedures were divided to simplify the reading. Also, as a side effect fixes SPARK-30049. 
What do you think?

**Comment by srowen** [comment]
Maybe @wangyum would be better to comment.
While I agree with the goal, this doesn't sever the connection to Hive. I'm also not clear why some Spark Scala code was rewritten here. How much are we sure the Hive code that was rewritten is equivalent and will remain so? maybe not a huge risk, but, I just don't know how much this buys vs the stated goal.

**Pull request closed without merging**